include/votca/xtp/matrixfreeoperator.h

```diff
@@ -0,0 +1,126 @@
+/* 
+ * Copyright 2009-2018 The VOTCA Development Team (http://www.votca.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef __VOTCA_TOOLS_MATRIX_FREE_OPERATOR_H
+#define __VOTCA_TOOLS_MATRIX_FREE_OPERATOR_H
+#include <votca/xtp/eigen.h>
+
+namespace votca { namespace xtp {
+
+    class MatrixFreeOperator : public Eigen::EigenBase<Eigen::MatrixXd>
+    {
+        public: 
+
+            typedef real_gwbse Scalar;
+            typedef real_gwbse RealScalar;
+            typedef int StorageIndex;
+
+            enum {
+                ColsAtCompileTime = Eigen::Dynamic,
+                MaxColsAtCompileTime = Eigen::Dynamic,
+                IsRowMajor = false
+            };
+
+            Index rows() const {return this-> _size;}
+            Index cols() const {return this-> _size;}
+
+            template<typename Vtype>
+            Eigen::Product<MatrixFreeOperator,Vtype,Eigen::AliasFreeProduct> operator*(const Eigen::MatrixBase<Vtype>& x) const {
+                return Eigen::Product<MatrixFreeOperator,Vtype,Eigen::AliasFreeProduct>(*this, x.derived());
+            }
+
+            // custom API
+            MatrixFreeOperator();
+
+            // convenience function
+            Eigen::MatrixXd get_full_matrix() const;
+            Eigen::VectorXd diagonal() const;
+            int size() const; 
+            void set_size(int size);
+
+            // extract row/col of the operator
+            virtual Eigen::VectorXd col(int index) const;
```

Comment:
make it a pure virtual function by adding"=0" then the compiler will throw an error if not implemented in dervived class